File: frontend/src/api/checkins.ts
import api from './client';

export interface CheckInRequest {
  instance_task_id: string;
  user_lat?: number;
  user_lng?: number;
  accuracy_meters?: number;
  photo_base64?: string;
  notes?: string;
  rating?: number;
}

export interface CheckInResponse {
  id: string;
  instance_task_id: string;
  verified: boolean;
  verified_by: string;
  reason: string;
  lat?: number;
  lng?: number;
  task_name: string;
  xp_earned: number;
  total_xp: number;
  level: number;
}

export interface ProgressDetail {
  instance_id: string;
  status: string;
  completed: number;
  total: number;
  progress_pct: number;
  xp_earned: number;
  xp_possible: number;
  tasks: {
    task_id: string;
    name: string;
    verification_type: string;
    xp: number;
    completed: boolean;
    verified_by?: string;
  }[];
}

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  username: string;
  display_name: string;
  total_xp: number;
  level: number;
}

export async function createCheckIn(req: CheckInRequest) {
  const { data } = await api.post<CheckInResponse>('/check-ins/', req);
  return data;
}

export async function getProgress(instanceId: string) {
  const { data } = await api.get<ProgressDetail>(`/check-ins/instance/${instanceId}/progress`);
  return data;
}

export async function getLeaderboard(limit = 20) {
  const { data } = await api.get<LeaderboardEntry[]>('/check-ins/leaderboard', { params: { limit } });
  return data;
}

File: app/schemas/checkins.py
"""Schemas for check-in endpoints."""
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class CheckInRequest(BaseModel):
    """Request to check in to an instance task."""
    
    instance_task_id: UUID = Field(..., description="ID of the instance task to check in to")
    user_id: UUID = Field(..., description="ID of the user (will come from auth later)")
    
    # GPS data (required for gps and both tasks)
    user_lat: Optional[float] = Field(None, description="User's current latitude")
    user_lng: Optional[float] = Field(None, description="User's current longitude")
    accuracy_meters: Optional[float] = Field(None, description="GPS accuracy in meters")
    
    # Photo data (required for photo and both tasks)
    photo_base64: Optional[str] = Field(None, description="Base64-encoded photo for verification")
    
    # Optional user input
    notes: Optional[str] = Field(None, max_length=1000)
    rating: Optional[int] = Field(None, ge=1, le=5, description="1-5 star rating")


class CheckInResponse(BaseModel):
    """Response after a check-in attempt."""
    
    id: UUID
    instance_task_id: UUID
    verified: bool
    verified_by: str
    reason: str
    
    # Location at check-in
    lat: Optional[Decimal] = None
    lng: Optional[Decimal] = None
    
    # Task info
    task_name: str
    
    class Config:
        from_attributes = True
